Implement xv6 Lottery Scheduler with Live Visualization

imyesung opened the pull request:
## Implementation Summary
- Implemented `settickets()` and `getpinfo()` system calls
- Added lottery scheduling algorithm to kernel
- Enabled ticket inheritance for child processes
- Created real-time distribution monitoring tool
- Built ASCII graph visualization for scheduler verification

## Test Results
- Tests 6 processes with 3:2:1 ticket ratio (30, 20, 10 tickets)
- Verified actual CPU distribution matches expected proportions
- Real-time graphs demonstrate scheduler fairness over time
- Final success verification shows proper lottery behavior

## Key Files Modified
- `kernel/proc.c`: Core lottery scheduler implementation
- `kernel/sysproc.c`: System call implementations
- `user/testlottery.c`: Testing and visualization tool
- `kernel/pstat.h`: Process statistics structure
- `user/user.h`: User-space system call declarations
- `kernel/syscall.c`: System call registration

## Testing Instructions
```bash
make qemu
$ testlottery
```

imyesung committed:
chore: Update .gitignore

imyesung committed:
feat: Add stacktrace system call for debugging

- kernel/syscall.h: define SYS_stacktrace (22)
- kernel/syscall.c: register handler
- kernel/sysproc.c: implement frame-pointer walk
- user/: add wrapper & test
- Makefile: include _stacktrace in UPROGS
- .gitignore: add macOS and IDE files

Enables runtime stack inspection by walking frame pointer
chain and collecting return addresses from user stack.

imyesung committed:
Add getreadcount() syscall for read() monitoring

Syscall 23 returns total read() calls since boot. Counter incremented
in sys_read(), simple int return. Test program shows usage example.

Why: helps understand system I/O behavior and debug read patterns.
How: global counter + syscall dispatch + user wrapper + test program.

imyesung committed:
Fix sys_read() argument validation order

Move read_count increment after successful argfd() validation to ensure
only valid read() calls are counted. This prevents counting failed calls
due to invalid file descriptors.

- Validate file descriptor first with argfd()
- Fetch buffer address and byte count via argaddr()/argint()
- Increment read_count only after validation succeeds
- Delegate to fileread() for the actual read operation

Follows xv6’s early-exit error-handling pattern.

imyesung committed:
Add lottery scheduler system calls infrastructure

Implement settickets() and getpinfo() syscalls for lottery scheduling.
Both processes start with 1 ticket by default, can be changed via settickets().

- Add pstat.h: defines process statistics structure for getpinfo()
- Add sys_settickets(): sets lottery tickets for calling process (min 1)
- Add sys_getpinfo(): fills pstat with all process info (tickets, ticks, etc)
- Add tickets/ticks fields to proc struct for lottery scheduler state
- Register syscalls 24/25 in syscall.h and syscall.c dispatch table

Why: enables proportional CPU scheduling based on ticket allocation.
How: kernel syscalls + process state tracking + user-space interface.

imyesung committed:
Implement lottery scheduler foundation with ticket inheritance

Add lottery scheduler infrastructure with process ticket management:
- Process initialization: all new processes start with 1 ticket (allocproc)
- Ticket inheritance: child processes inherit parent's tickets in fork()
- User space interface: add settickets/getpinfo declarations and stubs

Changes:
- kernel/proc.c: initialize p->tickets=1, p->ticks=0 in allocproc()
- kernel/proc.c: inherit tickets from parent in fork() (np->tickets = p->tickets)
- user/user.h: add settickets(int) and getpinfo(struct pstat*) declarations
- user/usys.pl: add assembly stubs for new system calls

TODO:
- [ ] Implement random number generator (kernel/rand.c)
- [ ] Replace round-robin scheduler with lottery algorithm in scheduler()
- [ ] Add tick counting when process gets scheduled
- [ ] Test with lottery scheduler test program
- [ ] Create performance graph showing 3:2:1 ticket ratio behavior
- [ ] Verify ticket inheritance works correctly across fork()

Status: System calls infrastructure complete, scheduler algorithm pending.

imyesung committed:
Fix kill() function loop boundary bug

Change for-loop bound from NCPU to NPROC in kill() function to properly
search through all process slots in the process table.

Bug: for(p = proc; p < &proc[NCPU]; p++) only checked first 8 processes
Fix: for(p = proc; p < &proc[NPROC]; p++) checks all 64 process slots

Why NPROC is correct:
- NCPU = number of CPU cores (typically 8)
- NPROC = maximum number of processes (typically 64)
- proc[] array has NPROC elements, not NCPU elements
- kill() needs to search ALL processes, not just first 8

Without this fix, kill() would fail to find processes with PID in slots
8-63, making those processes unkillable via kill() syscall.

imyesung committed:
Add pseudo-random number generator for lottery scheduler

Implement Linear Congruential Generator (LCG) for lottery scheduler:
- kernel/rand.c: randinit() and rand() functions using LCG algorithm
- kernel/defs.h: add randinit() and rand() function declarations
- kernel/main.c: initialize RNG during kernel boot with randinit()
- Makefile: add rand.o to kernel build objects

LCG parameters: a=1103515245, c=12345, returns 0-32767 range.
Simple but sufficient randomness for proportional CPU scheduling.

TODO:
- [ ] Replace round-robin scheduler() with lottery algorithm
- [ ] Increment process ticks when scheduled
- [ ] Test lottery scheduler with 3:2:1 ticket ratio
- [ ] Create performance graph showing proportional scheduling

Next: implement lottery algorithm in scheduler() function using rand().

imyesung committed:
Refactor random number generator to use standard srand() interface

Change from custom randinit() to standard srand() function:
- Replace randinit() with srand(unsigned int seed) for better compatibility
- Change unsigned long to unsigned int for consistency
- Remove randinit() call from main.c, move to procinit()
- Simplify comments and code structure

This provides a cleaner, more standard interface for the lottery scheduler.

imyesung committed:
Implement lottery scheduler algorithm (WIP - needs debugging)

Replace round-robin with proportional lottery scheduling:
- scheduler(): count total tickets, pick random winner, find owning process
- procinit(): initialize RNG with srand(1) for reproducibility
- settickets()/getpinfo(): complete system call implementations
- Process selection based on ticket proportions

Current Status: BUGGY - scheduler still behaves like round-robin
Test Results: All processes get 33% CPU time regardless of ticket ratio
Expected: 30:20:10 tickets should yield ~50%:33%:17% CPU distribution
Actual: A=33%, B=33%, C=33% (indicates lottery algorithm not working)

TODO: Debug why lottery selection isn't working properly
- Check if rand() is returning valid values
- Verify ticket counting logic in scheduler()
- Ensure proper process selection algorithm

imyesung committed:
Fix system call implementations and process structure

Update lottery scheduler system calls and process management:
- kernel/sysproc.c: clean up sys_getpinfo() with proper proc array access
- kernel/proc.h: ensure tickets/ticks fields are properly defined
- user/user.h: add missing struct pstat forward declaration

Fixes:
- Remove redundant extern declarations and variables
- Proper process table iteration in getpinfo()
- Clean user-space interface for lottery scheduler syscalls

imyesung committed:
Add lottery scheduler test program and build configuration

- user/testlottery.c: comprehensive test for lottery scheduler behavior
- Makefile: add _testlottery to UPROGS for building test program

Test features:
- Creates 3 child processes with 30:20:10 ticket ratio
- Measures CPU time distribution over multiple samples
- Displays results with ASCII bar chart and warning when proportional scheduling fails
- Expected: ~50%:33%:17% distribution, Actual: 33%:33%:33% (round-robin behavior)

Test confirms lottery scheduler is not working - all processes receive equal CPU time
regardless of ticket allocation, indicating the scheduler algorithm needs debugging.

imyesung committed:
config: Setup single CPU and test parameters

- Makefile: Change CPUS from 3 to 1 for deterministic testing
  Multi-core environment interferes with lottery scheduler behavior
- kernel/param.h: Add NSAMPLE=30 and SLEEP_TICKS=5 test parameters

Single CPU ensures accurate measurement of proportional scheduling.

imyesung committed:
kernel: Add lottery scheduler system calls

- kernel/rand.c: Replace srand() with randinit() for better randomization
- kernel/defs.h: Add settickets() and getpinfo() declarations
- kernel/sysproc.c: Implement lottery scheduler system call handlers

Enables user processes to control tickets and monitor scheduling behavior.

imyesung committed:
scheduler: Implement lottery algorithm

- kernel/proc.c: Replace round-robin with ticket-based lottery scheduling
- user/testlottery.c: Add error handling and optimize test workload

Processes with 30:20:10 tickets should get ~50%:33%:17% CPU time.

imyesung committed:
config: Enable multi-core environment for lottery scheduler

- Makefile: Change CPUS from 1 to 3 for proper multi-core testing
- Debug output shows lottery algorithm running on multiple CPUs
- Still shows 33%:33%:33% ratio indicating need for global coordination
- Next: implement proper multi-core lottery scheduling

imyesung committed:
fix(scheduler): serialize lottery scheduler on CPU0

- Resolve 'panic: release' caused by scheduler_lock mis-ordering
- Limit lottery scheduling to CPU0 to avoid multicore races
- Add startup delay in user/testlottery.c for deterministic init
- Preserve ticket-ratio correctness while restoring stability

TODO:
- Implement true multicore lottery scheduler with proper global locks
- Re-enable CPU1/CPU2 participation
- Distribute runnable set per-CPU for better cache locality

imyesung committed:
test: Increase children from 3 to 6 for multicore lottery testing

- Change NFORK to NCHILDREN with 6 processes (2x original)
- Duplicate ticket array: {30,20,10,30,20,10} to create CPU contention
- Add settickets(1) for parent to minimize interference
- Remove per-child creation messages to reduce output noise

Before: 3 processes on 3 CPUs = no real lottery competition
After: 6 processes on 3 CPUs = forces actual lottery scheduling

imyesung committed:
scheduler: Enable multicore lottery with global fairness counter

- Remove CPU0-only restriction, allow all CPUs to participate
- Add global_lottery_counter for system-wide fairness tracking
- Keep scheduler_lock from previous commit for serialization
- Change from 'simple multicore' to 'global fairness' approach

Before: Only CPU0 schedules (safe but inefficient)
After: All CPUs run lottery with shared state coordination

imyesung committed:
feat: Add process ticks tracking for lottery scheduler

- usertrap(): Track CPU time when timer interrupt occurs in user mode
- kerneltrap(): Track CPU time when timer interrupt occurs in kernel mode

imyesung committed:
test: Add lottery scheduler visualization with live graphs

- Display real-time CPU distribution across ticket groups
- Show ASCII bar graphs for 30/20/10 ticket ratios
- Track accumulated ticks and percentage distributions
- Visualize expected vs actual scheduling behavior
- Monitor 3:2:1 ratio producing proportional CPU allocation

imyesung committed:
test: Enhance lottery scheduler graphs with dual-bar visualization

- Replace simple graphs with detailed ASCII comparison charts
- Show actual (====) vs expected (----) distribution side-by-side
- Add 20-character bar graphs for clear percentage visualization
- Display tick counts and distribution ratios per monitoring round

imyesung commented:
@simnalamburt 리뷰 가능하신 때에 편히 리뷰 부탁드립니다 :)

simnalamburt commented on .gitignore:
```diff
@@ -15,3 +15,19 @@ mkfs
+
+# IDE
+.vscode/
+*.code-workspace
```
혼자 하시는 프로젝트라 큰 문제는 아니지만, 그래도 이런건 프로젝트의 `.gitignore`에 넣기보단 각자의 `git config core.excludesFile`로 관리하는게 좋은거같습니다.

https://git-scm.com/docs/git-config#Documentation/git-config.txt-coreexcludesFile

simnalamburt replied:
지금 .gitignore가 완전히 삭제되었는데 이건 좋지 않은거같습니다 project specific한 gitignore들은 있긴 해야해요

simnalamburt commented on kernel/param.h:
```diff
@@ -1,5 +1,5 @@
 #define NPROC        64  // maximum number of processes
-#define NCPU          8  // maximum number of CPUs
+#define NCPU          3  // maximum number of CPUs
```
NCPU를 8에서 3으로 줄이신 이유가 무엇인가요? 동시성 문제를 발견하려면 CPU 최대 숫자가 큰게 더 좋지않나요?

imyesung replied:
티켓 분배가 계속 균등하게 이루어지는 문제의 원인을 찾던 중에 `NCPU=1`로 테스트를 진행하였고
해당 문제가 멀티코어 환경에서만 발생함을 인지했습니다. `NCPU=8`이 컴퓨터 부하와 발열🫠 원인인가 싶어 멀티코어이되 적당히 낮은 `NCPU=3`으로 작업하던 중, NCPU 값보다 전체 프로세스 개수가 티켓 분배 스케줄링에 더 유의한 영향을 주는 것을 확인하고 나중에는 `NCPU=8`로 복구하여 진행했습니다.

simnalamburt commented on kernel/proc.c:
```diff
@@ -26,6 +27,29 @@ extern char trampoline[]; // trampoline.S
+ * GLOBAL LOTTERY STATE: Ensures fair lottery distribution across multiple CPUs
+ * by sharing lottery state and enforcing proper ticket proportions.
+ */
+static int global_lottery_counter = 0;  // Global counter for lottery fairness
```
이 scheduler_lock은 어떤 자원을 보호하기위해 존재하는 락인가요? 만약 보호하는 자원이 `global_lottery_counter` 단 한개라면, 글로벌 스케줄러 락을 없애시고 global_lottery_counter 를 atomic int로 만드시는쪽이 훨씬 효율적입니다. 그리고 global_lottery_counter 를 두지 말고, loterry_counter를 CPU별로 두거나, 아예 loterry_counter가 필요 없는 알고리즘으로 바꾸면 CPU들 사이에서 동기화를 할 필요가 없어져서 그쪽이 가장 빠릅니다.

imyesung replied:
기존 구조는 `scheduler_lock`을 통해 모든 CPU가 순차적으로 lottery를 수행하는 방식이었고,
`global_lottery_counter`로 전체 fairness를 조정하고 있었습니다.

[a19527b](https://github.com/imyesung/xv6-riscv/commit/a19527b26a1e9bad671a176d6ab595256a590265) 커밋에서는 `global_lottery_counter`를 제거해, 전역 락 없이도 각 CPU가 독립적으로 lottery scheduling을 수행하도록 리팩터링하였습니다.
- 각 CPU는 직접 snapshot을 만들고 winner를 선택
- 상태 전이(`RUNNABLE → RUNNING`)는 `p->lock`을 통해 직렬화

이로 인해 `scheduler_lock`이 더 이상 보호할 자원이 없어져, [3d1cbb4](https://github.com/imyesung/xv6-riscv/commit/3d1cbb43d5e228ab5e0c2e4de6d39e9cff42a18b) 커밋에서 해당 기능 및 선언을 제거하였습니다.

simnalamburt commented on kernel/sysfile.c:
```diff
@@ -72,10 +72,18 @@ sys_read(void)
+  argint(2, &n);
+
+  // Count valid read calls
+  read_count++;
```
`read_count`가 락이나 아토믹으로 보호되어있지 않아, 이러면 data race가 발생합니다.

imyesung replied:
data race 원인을 C→컴파일러→ISA→하드웨어 계층으로 다시 살펴보았고,
`read_count++`는 `__sync_fetch_and_add(&read_count, 1)`로 수정하였습니다. [cd3ac73](https://github.com/imyesung/xv6-riscv/commit/cd3ac73ec73c794fb31f025fc0e818a79b39b295)

simnalamburt commented on kernel/proc.c:
```diff
@@ -26,6 +27,29 @@ extern char trampoline[]; // trampoline.S
+static int global_lottery_counter = 0;  // Global counter for lottery fairness
+
+// External declaration - defined in main.c
+extern int read_count;
```
defs.h 에서 이미 `extern int read_count;` 를 선언했기때문에 여기에서 다시 할 필요가 없습니다.

simnalamburt commented on kernel/sysproc.c:
```diff
@@ -91,3 +95,97 @@ sys_uptime(void)
   return xticks;
 }
+
+extern int read_count;  // Direct access to global counter
```
defs.h 에서 이미 `extern int read_count;` 를 선언했기때문에 여기에서 다시 할 필요가 없습니다.

imyesung replied:
`read_count`는 `defs.h`의 선언만 참조하도록 변경했습니다. [98600e5](https://github.com/imyesung/xv6-riscv/commit/98600e55f409d6bdec3ce78b38e1f571011ee97a)

simnalamburt commented on kernel/trap.c:
```diff
@@ -77,8 +77,14 @@ usertrap(void)
+    struct proc *p = myproc();
+    if(p != 0) {
+        p->ticks++;
+    }
```
주석의 설명과는 다르게, Lottery Scheduling을 구현하기위해 CPU time (ticks) 를 추적할 필요가 없습니다. 지금 ticks는 순수히 통계적 목적으로 만드신걸로 보이는데, 그렇다면 주석의 설명도 그렇게 쓰셔야합니다.

그리고 48번 라인에서 이미 `struct proc *p = myproc();` 를 호출했기때문에 `myproc()` 함수를 다시 호출하지 말고 위의 결과를 재활용하는것이 더 효율적입니다.

imyesung replied:
ticks의 의미를 기존 'timer interrupt 기반 실행 시간 누적'에서 '디스패치 횟수 기반 스케줄링 카운터'로 명확히 변경하였고 향후 procdump에서 fairness check(실측 분포 vs 티켓 비율) 지표로 활용할 예정입니다.

[11ac9ac](https://github.com/imyesung/xv6-riscv/commit/11ac9ac56e002557cdc1b1576d30710cb9b66a50) 커밋에서
- `usertrap()` 및 `kerneltrap()`에서의 timer interrupt 기반 ticks 증가를 제거
- `scheduler()` 내 프로세스가 `RUNNABLE` → `RUNNING` 상태로 전이되는 디스패치 시점에서만 ticks++ 발생하도록 수정
- 관련 주석 또한 해당 의미에 맞게 재정비했습니다.

simnalamburt commented on kernel/proc.c:
```diff
@@ -446,36 +490,310 @@ scheduler(void)
-        // to release its lock and then reacquire it
-        // before jumping back to us.
         p->state = RUNNING;
+        p->ticks++;
```
이 위치에서 ticks를 증가시키는 이유가 무엇인가요?

simnalamburt commented on kernel/proc.c:
```diff
@@ -446,36 +490,310 @@ scheduler(void)
+          //          cpuid(), p->pid, p->name, p->tickets);
+          // }
+          p->state = RUNNING;
+          p->ticks++;              // Increment runtime counter for statistics
```
이 위치에서 ticks를 증가시키는 이유가 무엇인가요?

simnalamburt commented on kernel/trap.c:
```diff
@@ -151,8 +157,10 @@ kerneltrap()
   // give up the CPU if this is a timer interrupt.
-  if(which_dev == 2 && myproc() != 0)
+  if(which_dev == 2 && myproc() != 0) {
+    myproc()->ticks++;
```
`myproc()` 함수를 두번 호출하셨는데, 비효율적일뿐만 아니라 앞의 함수 호출 결과가 뒤의 함수 호출 결과와 같다는 보장이 없습니다

imyesung replied:
[acbabdc](https://github.com/imyesung/xv6-riscv/commit/acbabdca2cf36f35ddafa520d3169f7ea1d8853d) 커밋에서 함수 호출 관련 수정을 진행하였으나,
[11ac9ac](https://github.com/imyesung/xv6-riscv/commit/11ac9ac56e002557cdc1b1576d30710cb9b66a50) 이후 `usertrap()`의 timer 기반 ticks++는 디스패치 카운터로 재정의되어 더 이상 의미가 없는 부분 확인했습니다.
후속 커밋에서 해당 증가 코드를 제거하고, `myproc()`은 1회 호출 후 재사용하도록 정리하겠습니다.

simnalamburt commented on kernel/proc.c:
```diff
@@ -276,6 +315,8 @@ growproc(int n)
 
 // Create a new process, copying the parent.
 // Sets up child kernel stack to return as if from fork() system call.
+// fork() 함수 전체를 아래 코드로 교체하세요.
```
불필요한 주석

simnalamburt commented on kernel/proc.c:
```diff
@@ -312,14 +356,14 @@ fork(void)
   np->state = RUNNABLE;
+  
+  // Finally, release the lock.
   release(&np->lock);
```
이부분 중간의 `release(&np->lock);`, `acquire(&np->lock);` 가 삭제되었는데 데드락 소지가 있어보입니다. 이 수정의 이유는 무엇인가요?

imyesung replied:
의도는 락 전환 최소화였으나, `np->lock` 보유 상태에서 `wait_lock`을 획득하려는 경로가 생겨 교착 소지가 있는 부분 확인했습니다.
[e1b5505](https://github.com/imyesung/xv6-riscv/commit/e1b550586db83e996618c1a2aad40556904460e3) 커밋에서 복원하였습니다.

simnalamburt commented on kernel/rand.c:
```diff
@@ -0,0 +1,27 @@
+randinit(void)
+{
+  initlock(&rand_lock, "rand");
+}
```
랜덤시드가 1로 고정되어있어서 이러면 항상 같은결과가 나옵니다. 랜덤시드를 고정이 아닌 값을 주면 좋을것같습니다.

더불어 rand() 함수는 스케줄러가 매번 호출하기때문에 상당히 빨라야하는데, 글로벌 락이 있으면 안좋습니다. 어차피 여러 CPU가 하나의 rand state를 공유할 필요는 없는데, state를 CPU마다 두는식으로 완전히 병렬 실행이 가능한 구조로 바꾸는걸 고민해보시면 좋을것같습니다.

simnalamburt commented on kernel/syscall.c:
```diff
@@ -112,20 +116,24 @@ static uint64 (*syscalls[])(void) = {
-[SYS_unlink]  sys_unlink,
-[SYS_link]    sys_link,
-[SYS_mkdir]   sys_mkdir,
 [SYS_close]   sys_close,
```
크게 중요한건 아니지만 기존 함수의 위치를 옮길 필요가 있나 싶습니다. diff가 불필요하게 커져서요

imyesung replied:
추후 작업에서는 기능 변경과 위치 변경을 분리해 diff 노이즈를 줄이겠습니다. 감사합니다.

simnalamburt commented on kernel/proc.c:
```diff
@@ -446,36 +490,310 @@ scheduler(void)
+     * favored while maintaining the correct proportional distribution.
+     */
+    global_lottery_counter++;
+    int winner = (rand() + r_time() + cpuid() + global_lottery_counter) % total_tickets;
```
어차피 rand() 결과가 충분히 랜덤한데 `+ r_time() + cpuid() + global_lottery_counter` 를 할 필요가 있나 싶습니다.

그리고 `% total_tickets` 는 [0, 32768) 범위의 랜덤 결과를 이걸 [0, total_tickets) 범위로 바꾸시려고 이렇게 하신거같은데, 이러면 uniform하지 않아집니다. 아래 링크를 한번 봐보시고, 어떻게하면 uniform을 유지하면서 범위를 줄일 수 있는지 고민하고 고쳐보셔요.

<https://github.com/simnalamburt/snippets/blob/c1d19d7b19/python/uniformness-of-random/main.ipynb?rgh-link-date=2025-07-22T13%3A33%3A10Z>

imyesung replied:
[395d2f0](https://github.com/imyesung/xv6-riscv/commit/395d2f0988f89f3c90a0f6ea136e6a5079270c10) 커밋에서 리팩터링 진행하였으나
`rand.c`를 살펴보면서 modulo bias, per-CPU PRNG 설계 등 기본기가 부족함을 인지했습니다.
첨부 자료를 참고해 기초를 보완한 뒤 후속 PR로 보강하겠습니다.

simnalamburt commented on kernel/sysproc.c:
```diff
@@ -91,3 +95,97 @@ sys_uptime(void)
+  int tickets;
+  argint(0, &tickets);
+  
+  return settickets(tickets);
```
사용자가 tickets로 이상한값(0, 음수)을 입력하지는 않았는지 검사해야합니다

imyesung replied:
[d65cdc2](https://github.com/imyesung/xv6-riscv/commit/d65cdc280477fae9c8da11c5c11de1dea167c55c) 커밋으로 유저 입력값 검증 추가하였습니다. 감사합니다.

simnalamburt commented on kernel/proc.h:
```diff
@@ -104,4 +104,8 @@ struct proc {
+  
+  // Lottery scheduler fields
+  int tickets;                 // Number of lottery tickets
+  int ticks;                   // Number of times scheduled
```
둘 다 항상 양수여야하는값인데 uint를 쓰면 어떨까 싶습니다.

그리고 ticks의 경우, 타이머 인터럽트당 1씩 오르는 값이고, xv6-riscv 기본설정이 1초에 타이머인터럽트가 10회씩 발생하게되어있어서 서버를 (2\*\*31 - 1/10)초 ≈ 6.8년 켜둘경우 ticks에서 오버플로우가 발생합니다. 유닉스 서버는 오래 켜두는일이 많아서 6.8년이 그다지 큰 값이 아닙니다. uint64로 바꾸면 어떨까 싶습니다. uint64 로 하면 (2**64 - 1)/10초 ≈ 58454204609년 ≈ 우주나이를 4.24번 반복해야 오버플로우가 발생하므로 안전하다고 할 수 있습니다.

imyesung replied:
tickets: 음수 불가하도록 uint
ticks: 안정성을 고려하여 uint64로 수정하였습니다.
[2377987](https://github.com/imyesung/xv6-riscv/commit/2377987ebfc215a41b4033e356d7a4f0143a239c)

simnalamburt commented on kernel/rand.c:
```diff
@@ -0,0 +1,27 @@
+#include "riscv.h"
+#include "spinlock.h"
+#include "proc.h"
+#include "defs.h"
```
정말로 필요한것만 include 해주시는것이 좋습니다

imyesung replied:
평소 체크를 더 꼼꼼히 하고, CI에서의 Static checks도 공부해보겠습니다!

simnalamburt commented on kernel/pstat.h:
```diff
@@ -0,0 +1,13 @@
+  int inuse[NPROC];   // whether this slot of the process table is in use (1 or 0)
+  int tickets[NPROC]; // the number of tickets this process has
+  int pid[NPROC];     // the PID of each process 
+  int ticks[NPROC];   // the number of ticks each process has accumulated 
```
이것도 int여야할지 uint여야할지 uint64여야할지 한번 재고해주시면 좋을것같습니다

imyesung replied:
- tickets: 음수 불가하도록 uint
- ticks: 안정성을 고려하여 uint64로 수정하였습니다.
[2377987](https://github.com/imyesung/xv6-riscv/commit/2377987ebfc215a41b4033e356d7a4f0143a239c)

simnalamburt commented on kernel/sysproc.c:
```diff
@@ -91,3 +95,97 @@ sys_uptime(void)
+
+extern int read_count;  // Direct access to global counter
+
+uint64
+sys_getreadcount(void)
+{
+  return read_count;  // Simple and direct
+}
```
`read_count`는 int인데 `sys_getreadcount`의 리턴값은 uint64여서 차이가 있습니다.

imyesung replied:
[3918460](https://github.com/imyesung/xv6-riscv/commit/39184601c905752155601f7b9b50230a813e94d6)
`read_count`: int -> uint64
시스템콜 리턴 타입과 동일하게 수정하였습니다.

simnalamburt commented on user/getreadcount.c:
```diff
@@ -0,0 +1,29 @@
+int
+main(int argc, char *argv[])
+{
+  int before, after;
```
사소한거지만 변수선언을 함수 맨 앞에 몰아서 하실 필요가 없습니다. 가능한한 변수를 나중에 선언하는것이 일반적으로 좋은 practice입니다.

imyesung replied:
초기화 전 읽기 가능성을 줄이기 위해 함수 시작부의 일괄 변수 선언을 제거하고,
변수를 사용 지점에서 선언하도록 스코프를 축소했습니다. 동작 변화는 없습니다.
[08f5ed4](https://github.com/imyesung/xv6-riscv/commit/08f5ed41f5ad84f09278cb8c4783059aaa9b7775)

simnalamburt commented on user/testlottery.c:
잘 짜신걸로 보입니다만 그래프 결과도 같이 첨부해주시면 좋을것같습니다. 그리고 그래프를 굳이 C 코드로 그리실필요는 없고, C 코드로는 표만 출력하고 실제 그래프 그리기는 엑셀이나 파이썬 쓰시는게 편하실거같아요.

imyesung replied:
<img width="523" height="396" alt="xv6-lottery-scheduler-round20" src="https://github.com/user-attachments/assets/d3002a6c-6992-47fe-ab8b-b18580ba571a" />

[209de48](https://github.com/imyesung/xv6-riscv/commit/209de483a5d465808dd289c89bb2973a6324db89)에 추가하였습니다.
다만 소수점 단위를 절사하고, 해당 부분 명확히 코드에서도 명시하지 않아 티켓 분포가 총 98%인 현상이 있습니다.항상 결과가 100%로 나오도록 Largest Remainder 적용해볼 예정입니다.

simnalamburt replied:
그래프가 무슨뜻인지 잘 이해가 안갑니당

simnalamburt commented on kernel/proc.c:
```diff
@@ -446,36 +490,310 @@ scheduler(void)
+     */
+    if(scheduler_lock_held) {
+      release(&scheduler_lock);
     }
```
주석을 상세히 달아주신건 좋지만 코드가 한눈에 들어오지 않아서, 주석을 큰 단위별로 몰아서 달아주시면 좋을것같습니다.

imyesung replied:
좋은 주석의 역할과 코드란 뭘지 고민하면서 계속 작업해보겠습니다!

simnalamburt reviewed (changes requested):
그래프도 함께 첨부해주시면 좋을거같습니다

simnalamburt commented:
아 그리고 다음에는 여러 과제를 한번에 풀더라도 과제별로 분리해서 PR 날려주세용

imyesung committed:
atomic: Replace read_count++ with __sync_fetch_and_add

- Replaced non-atomic increment (read_count++) with atomic operation (__sync_fetch_and_add)
- Resolves potential race conditions when multiple processes invoke sys_read concurrently
- Addresses reviewer feedback regarding atomic operations and thread safety

imyesung committed:
fix: Prevent deadlock by adjusting lock acquisition order in fork()

- Release np->lock temporarily before acquiring wait_lock
- Set parent process pointer safely under wait_lock protection
- Re-acquire np->lock before setting RUNNABLE state
- Addresses potential deadlock issue highlighted in code review

imyesung committed:
validation: Reject invalid ticket values in sys_settickets()

- Reject ticket values less than 1 (zero, negative numbers)
- Return -1 error code upon invalid input
- Addresses reviewer feedback on input sanitization

imyesung committed:
fix: Remove duplicate read_count definition in sysfile.c

- Changed to 'extern int read_count;' to avoid multiple definition error
- Prevents build failure when linking sysfile.c and sysproc.c

imyesung committed:
rand: Refactor to per-CPU RNG and improve winner selection

- Replaced global random state with per-CPU seed and lock (next[NCPU], rand_lock[NCPU])
- Prevents contention and improves scalability in rand()
- Changed winner selection to use uniform scaling: (rand() * total_tickets) / 32768
- Ensures fair and unbiased lottery scheduling outcome
- Removed unused includes from rand.c for clarity

imyesung committed:
cleanup: Remove redundant 'extern int read_count' declarations

- Deleted duplicate extern declarations in proc.c and sysproc.c (already in defs.h)

imyesung committed:
refactor: Simplify timer interrupt logic in usertrap()

- Call myproc() once and reuse return value
- Reordered p->ticks++ before yield() for logical clarity
- Tick semantics unchanged

imyesung committed:
fix: Change read_count type to uint64 for syscall consistency

- Update read_count type from int to uint64 in defs.h and main.c
- Match syscall return type in sys_getreadcount()
- Remove redundant extern declarations in sysfile.c

imyesung committed:
refactor: Change scheduler stat fields to uint/uint64

- Change proc.tickets to uint and ticks to uint64
- Update pstat structure accordingly
- Add types.h includes where necessary

imyesung committed:
revert(param): Restore NCPU to 8 for multicore baseline

- Revert a local single-core experiment (3) and restore the default multicore value (8)
- Keep reproducibility and consistent behavior under multicore scheduling tests

imyesung committed:
refactor(scheduler): Drop global lottery state; prefer per-CPU selection

- Remove global_lottery_counter and inline comments to eliminate shared state
- Reduce lock contention and serialization points across CPUs
- Keep scheduler_lock temporarily; plan follow-up removal if it protects nothing

imyesung committed:
test: Report convergence delta and ensure child cleanup

- Print delta versus the expected 50:33:17 ratio to quantify convergence
- Kill and wait all child processes to guarantee clean teardown and stable metrics

imyesung committed:
rand: Replace LCG with per-CPU PCG32

Switch from legacy LCG to PCG32 (XSH-RR) algorithm.
- Per-CPU state eliminates locks
- Seed with splitmix64(time + cpuid)
- Keep xv6 API: rand() returns 0-32767
- Better statistical quality; no contention on the RNG path

imyesung committed:
scheduler: Remove global lock for lottery

Each CPU runs lottery independently.
- No more global scheduler_lock
- Scan runnable procs, draw winner locally
- Switch under p->lock only
- Note: uses rand()%total (has modulo bias)

imyesung committed:
test: Restructure lottery test into 3 ticket groups

Create 3 workers per ticket group {30,20,10} → 9 workers total
to improve statistical stability. Fix total ticket reporting
(180 from workers + noise) and print noise ratio. Keep tick-based
sleeps for fast runs. Reduce noise processes and increase work
iterations to amplify the signal.

imyesung committed:
scheduler: Define per-proc ticks as dispatch count

Clarify semantics: per-proc 'ticks' is a dispatch counter.
- Increment once when a process is selected RUNNING
- Remove timer-based increments in usertrap/kerneltrap
- Update proc.h comment for unambiguous meaning
This avoids double counting and keeps lottery stats consistent.

imyesung committed:
proc: Make procdump read-consistent

Acquire p->lock briefly to read state/pid/name, then release before printing.
Early-continue on UNUSED. Copy name locally. Avoid printing under lock.

imyesung committed:
chore(git): Remove project-level .gitignore

Move personal environment patterns to user's global gitignore.
Keep repo-level rules focused on project-wide artifacts only.

imyesung committed:
api(syscall): Make getreadcount return uint64 across kernel/user

Unify prototype with sys_getreadcount(uint64) and prevent truncation.
- kernel/defs.h: int -> uint64
- user/user.h  : int -> uint64
This aligns the public interface and avoids future overflow issues.

imyesung committed:
refactor(pstat): Use int fields; cap downcasts in getpinfo()

Simplify user-visible pstat to int and guard downcasts.
- pstat.{tickets,pid,ticks}: int
- getpinfo(): clamp uint/uint64 to INT_MAX before copyout
- types.h: define INT_MAX for clamping
Notes: pstat is a user ABI; update user tools accordingly.

imyesung committed:
fix(user): Print uint64 with %lu and proper casts

Match printf format with uint64(=unsigned long) to satisfy -Werror=format.
- before/after counters printed via %lu with casts

imyesung committed:
merge: Integrate fix-lottery-feedback into review-lottery

imyesung committed:
docs(lottery): Add scheduler snapshot for round 20

imyesung committed:
docs(readme): Rename to README.md for GitHub Markdown rendering

imyesung commented:
@simnalamburt 방금 re-review 요청 드렸습니다.
[merge commit](https://github.com/imyesung/xv6-riscv/commit/2d1e00725a8798f2f0f761cd3f16bfc9644595d2)에서 리뷰 피드백 반영분을 머지했습니다.
편히 확인 부탁드립니다. 감사합니다.

simnalamburt commented on kernel/trap.c:
```diff
@@ -77,8 +77,12 @@ usertrap(void)
   // give up the CPU if this is a timer interrupt.
-  if(which_dev == 2)
-    yield();
+  if(which_dev == 2) {
+    struct proc *p = myproc();
+    if(p != 0) {
+      yield();
+    }
+  }
```
48번 라인에서 이미 `myproc()` 을 호출하고 그 결과를 변수로 저장하고있어 여기서 다시 `myproc()`을 호출할 필요가 없습니다.

아마 기존 `myproc()` 호출을 재활용하도록 수정하고나면 이 파일은 diff가 완전히 없어질것같네요.

simnalamburt commented on kernel/trap.c:
```diff
@@ -151,8 +155,9 @@ kerneltrap()
 
   // give up the CPU if this is a timer interrupt.
-  if(which_dev == 2 && myproc() != 0)
+  if(which_dev == 2 && myproc() != 0) {
     yield();
+  }
```
이전 코드와 비교해 중괄호 추가 외에는 달라진게 없어보여서 이런거는 이전과 완전히 동일하게 만들어서 diff를 제거하면 더 좋습니다

simnalamburt commented on kernel/proc.c:
```diff
@@ -670,26 +714,87 @@ either_copyin(void *dst, int user_src, uint64 src, uint64 len)
+{
+  struct proc *p = myproc();
+
+  if(tickets < 0)
+    return -1;
```
tickets가 0인것을 허용하면 안되기때문에 `tickets < 1` 를 체크해야합니다. 그리고 `sys_settickets()` 에서 이미 티켓 수 검증을 하고있어서 여기에서 다시 할 필요는 없습니다

simnalamburt commented on kernel/rand.c:
```diff
@@ -0,0 +1,49 @@
+#include "riscv.h"
+#include "defs.h"
+
+// PCG32 state per CPU
+static uint64 pcg_state[NCPU];
+static uint64 pcg_inc[NCPU];   // stream selection (must be odd)
+
+static inline uint32 
+rotl32(uint32 x, unsigned r) {
+    return (x << r) | (x >> (32 - r));
+}
+
+static inline uint32 
+pcg32_step(int cpu) {
+    uint64 oldstate = pcg_state[cpu];
+    pcg_state[cpu] = oldstate * 6364136223846793005ULL + pcg_inc[cpu];
+    
+    // Output function (XSH RR): xorshift high 32 bits and rotate right
+    uint32 xorshifted = (uint32)(((oldstate >> 18u) ^ oldstate) >> 27u);
+    uint32 rot = (uint32)(oldstate >> 59u);
+    return rotl32(xorshifted, rot);
+}
```
랜덤 함수에서 글로벌 동기화를 피하기위해 CPU별로 랜덤 state를 따로 만드신것 좋습니다 👍 

다만 Per-CPU state가 한곳에 모여있는게 코드 가독성/유지보수성이 더 좋아서, @Klassikcat 님께서 하신것처럼 `kernel/proc.h` 의 `struct cpu` 에 state를 추가하는게 더 좋지않을까 싶습니다.

더불어 pcg_inc 는 변하는값이 아니라 매우 빠르게 계산할 수 있는 상수여서, 변수로 저장하지말고 수식 안에 넣어버리는게 어떨까 싶습니다.

###### References
- https://github.com/Klassikcat/xv6-riscv/compare/88cc711..2a9a8a4#diff-3e06d244cbfbd7a5db4cc7dabc8a9b07e38ef4a345687c6bce70ef4f40268a38

simnalamburt commented on kernel/proc.c:
```diff
@@ -446,37 +460,67 @@ scheduler(void)
+      continue;  // idle - no RUNNABLE processes
     }
+
+    // randomly select a winner based on the total tickets
+    uint winner = (rand() % total) + 1;
```
`% total` 이 아직 남아있는데, @baksohyeon 님께서 하신것처럼 https://github.com/baksohyeon/xv6-riscv/pull/4#discussion_r2233668346 이렇게 하시면 됩니다. 다만 제가 저기에도 쓴것처럼 범위를 줄이는 동작을 rand() 에서 하기보단 rand 함수에 파라미터를 추가해 rand 함수가 범위를 줄이는 동작을 제공하도록 하는게 더 좋아보입니다.

simnalamburt commented on kernel/pstat.h:
```diff
@@ -0,0 +1,14 @@
+
+struct pstat {
+  int inuse[NPROC];   // whether this slot of the process table is in use (1 or 0)
+  int tickets[NPROC]; // the number of tickets this process has
+  int pid[NPROC];     // the PID of each process
+  int ticks[NPROC];   // the number of ticks each process has accumulated
```
`struct proc`의 tickets/ticks가 uint/uint64여서 여기에서도 같은 타입을 써줘야합니다.

살펴보니 정상적으로 수정하셨다가 https://github.com/imyesung/xv6-riscv/commit/78fe2deffae477c2666c4fda715d9dc853ae8ec0 이 커밋에서 다시 int로 바꾸신것같은데 이유가 무엇인가요?

simnalamburt commented on kernel/sysfile.c:
```diff
@@ -72,10 +75,18 @@ sys_read(void)
+  argaddr(1, &p);
+  argint(2, &n);
+
+  // Count valid read calls using atomic operation
+  __sync_fetch_and_add(&read_count, 1);
```
이렇게 하면 데이터레이스가 사라지기는 하니, 이것도 맞는 해결책인것은 맞습니다. 다만 __sync_fetch_and_add 는 비표준 gcc 확장이어서 read_count 전역변수를 _Atomic 변수로 만드는게 표준 C 만 사용하는 더 좋은 코드이긴 합니다. 아래 레퍼런스를 참고해보셔요

###### Reference
- https://en.cppreference.com/w/c/language/atomic.html

simnalamburt commented on .gitignore:
개인적인 gitignore 설정들을 gitexclude로 옮긴것은 좋긴한데, 기존 .gitignore 설정은 사람에 따라 차이 없이 다들 공유하여 써야하는 설정이라 기존 gitignore를 완전히 삭제하는건 좋지 않은것같습니다.

simnalamburt commented on kernel/sysfile.c:
```diff
@@ -16,6 +16,9 @@
 #include "file.h"
 #include "fcntl.h"
 
+// Global counter for read system calls
+extern uint64 read_count;
+
```
`defs.h` 에 이미 `extern uint64 read_count` 가 선언되어있어서 이 문장은 필요없습니다.

simnalamburt commented on kernel/proc.c:
```diff
@@ -446,37 +460,67 @@ scheduler(void)
-        // It should have changed its p->state before coming back.
-        c->proc = 0;
-        found = 1;
+        total += (p->tickets > 0) ? p->tickets : 1;
```
`(p->tickets > 0) ? p->tickets : 1` 를 매번 검사하기보다는 p->tickets 가 항상 양수가 되도록 하는게 더 좋은 패턴인데, 지금 이미 sys_settickets 에서 그렇게 검사를 하고있어서 항상 양수임이 보장됩니다.

simnalamburt commented on kernel/proc.c:
```diff
@@ -446,37 +460,67 @@ scheduler(void)
+    }
+    release(&picked->lock);
+
+    // Give other CPUs a chance
+    #ifdef CFS
+    yield();
+    #endif
```
`#ifdef CFS`는 무슨 의미인가요?

simnalamburt commented on kernel/proc.c:
```diff
@@ -446,37 +460,67 @@ scheduler(void)
       }
       release(&p->lock);
     }
-    if(found == 0) {
-      // nothing to run; stop running on this core until an interrupt.
-      intr_on();
-      asm volatile("wfi");
```
기존의 `intr_on(); asm volatile("wfi");` 이 코드가 실행할 프로세스가 없을때 CPU를 쉬게하는 명령어였는데, 이게 삭제되고 busy wait으로 대체되어서 이러면 CPU를 항상 100%로 실행하게됩니다.
